Anlab.Mvc/Controllers/PaymentController.cs

```diff
@@ -0,0 +1,337 @@
+
+            // check signature
+            var dictionary = Request.Form.ToDictionary(x => x.Key.ToString(), x => x.Value.ToString());
+            if (!_dataSigningService.Check(dictionary, response.Signature))
+            {
+                Log.Error("Check Signature Failure");
+                ErrorMessage = "An error has occurred. Payment not processed. If you experience further problems, contact us.";
+                return RedirectToAction("Index", "Home");
+            }
+
+            //ProcessPaymentEvent(response, dictionary); //For testing locally, can enable this.
+
+            var order = _context.Orders.SingleOrDefault(a => a.Id == response.Req_Reference_Number);
+            if (order == null)
+            {
+                Log.Error("Order not found {0}", response.Req_Reference_Number);
+                ErrorMessage = "Order for payment not found. Please contact technical support.";
+                return NotFound(response.Req_Reference_Number);
+            }
+            if (order.CreatorId != CurrentUserId)
+            {
+                ErrorMessage = "You don't have access to this order.";
+                return NotFound(response.Req_Reference_Number);
+            }
+
+            var responseValid = CheckResponse(response);
+            if (!responseValid.IsValid)
+            {
+                ErrorMessage = ErrorMessage = string.Format("Errors detected: {0}", string.Join(",", responseValid.Errors));
+                return RedirectToAction("Pay", new {id = order.Id});
+            }
+
+            //Should be good,   
+            if (order.Status == OrderStatusCodes.AwaitingPayment)
+            {
+                order.Status = OrderStatusCodes.Paid;
+                _context.SaveChanges(true);
+            }
+            Message = "Payment Processed. Thank You.";
+
+            //ViewBag.PaymentDictionary = dictionary; //Debugging. Remove when not needed
+
+            return View(response);
+        }
+
+        [HttpPost]
+        [IgnoreAntiforgeryToken]
+        public ActionResult Cancel(ReceiptResponseModel response)
+        {
+            {
+                Log.ForContext("response", response, true).Information("Cancel response received");
+
+                // check signature
+                var dictionary = Request.Form.ToDictionary(x => x.Key.ToString(), x => x.Value.ToString());
+                if (!_dataSigningService.Check(dictionary, response.Signature))
+                {
+                    Log.Error("Check Signature Failure");
+                    ErrorMessage = "An error has occurred. Payment not processed. If you experience further problems, contact us.";
+                    return RedirectToAction("Index", "Home");
+                }
+                //ViewBag.PaymentDictionary = dictionary; //Debugging. Remove when not needed
+                ProcessPaymentEvent(response, dictionary); //TODO: Do we want to try to write cancel events?

```

Comment:
@jSylvestre if we do we'll have to be careful because we'll want to find the payment even associated w/ an order.  If there are multiple then it could get strange if we don't pay attention